### ISSUE TITLE
Improve content selection in Printing

### ISSUE THREAD

**Issue by fvisin** [bartvm/blocks#520]
`Printing` should be more configurable in choosing what it actually prints. The current leading underscore logic to filter out channels was meant to be a short-term solution and is not always optimal. It would be nice to be able to select (or ignore) channels specifying a list of channels names or regex rules.

**Comment by bartvm** [comment]
Mm... Is that really necessary though? I agree that the underscores aren't a good solution, and I'm planning on changing that to an `exclude`/`hidden` list of values that won't be printed by default, which extensions and/or the user can edit.

However, matching regular expressions etc. sounds like overkill to me. If you want to do something that complicated, then log to a database (which will soon be possible) and run queries on that, or pipe your output to a text file on which you run `grep`.

I'm not saying it's not a useful function, but as I pointed out in #517; I think we need to start paying attention to avoiding feature bloat considering the rate at which the codebase is growing, and I would argue that this is a good example of that.

**Comment by fvisin** [comment]
I agree that matching regular expressions could be overkill, but on the other hand it is something you already do in `VariableFilter`. I don't think that implementing it in `Printing` too would be very time consuming or require much code to write/maintain. 

For what concerns filtering in general, I think it is really necessary and it goes beyond the analysis of the data (that you could do logging to a database or to an external file): Bricks access the monitored channels to get the data they need, so sometimes you might want to add variables to the DataStreamMonitoring but you don't want to display them in the log.

For instance, I am Plotting real time data with Bokeh and to do that I add the variables I need to the monitored channels. I plot that data because it's not readable: it doesn't make any sense to display it in the textual log and it makes the (useful) log information impossible to read.

**Comment by dwf** [comment]
While I share your concerns @bartvm, I think it's one of the complaints I
heard most often from pylearn2 novices. We've solved that partly by being a
little more selective about the default channels that get added, but I
think having some way forward on this issue is probably a good thing for
core to have. I'm not saying regexes are necessarily the way (and the
overhead of running the regex engine every time something gets printed
might become worrisome if printing is frequent enough), but it would be
nice if there were something short of running your script piped to grep -v.

**Comment by fvisin** [comment]
To clarify, I didn't want to say that using regular expression is the way to go. It might be a powerful solution, but even just a list of included or excluded variables could work. 
Maybe something like this:
* __Printing(show_channels = None, hide_channels = None):__ all the variables get printed
* __Printing(show_channels = [a,b,c], hide_channels = None):__ only `a`, `b` and `c` get printed
* __Printing(show_channels = None, hide_channels = [`a`,`b`]) :__ all the variables apart from `a` and `b` get printed

**Comment by rizar** [comment]
A very lean and mean way to support output filtering would to be to have a
callback argument in Printing. The callback can return False if a
particular log record should not be printed. The user would receive full
control and could decide himself whether he prefers a faster or simpler
implementation.

**Comment by bartvm** [comment]
Well, at the very least I think we should postpone this option until after we finally have something that's worthy of Blocks 0.1, and maybe it's something that could go in `blocks-contrib` instead.

I know it was a common complaint for Pylearn2, but I think the issue of having too many channels printed to your screen is actually more indicative of the fact that simply dumping your log to the screen each batch/epoch is a bad way of monitoring your training progress. I'd rather see three ways of monitoring:

* Dump your log, all of it. Quick and dirty, simple to code, and can be piped to a text file if you want to. Basically what we have now.
* Some kind of minimal `ncurses` interface with e.g. two columns: Monitored channels in the left column, and the (iterations_done, value) pairs in the right column.
* Plotting through a web interface (like Bokeh)

Seems like an `ncurses` interface just makes things more complicated, but I think it's a more sensible solution than all of these pattern matching options. They all look nice, but they don't scale well at all. Once you've got 100 channels and want to match 50, you still end up writing a long list of patterns/regular expressions/etc. In those cases I'd prefer having a super-simple interface where I can select the record I care about (maybe even search for it) and get a column with the latest values (makes it much easier to compare anyway than having to compare values 100 lines across).

This would be a great `blocks-contrib` extension.

**Comment by rizar** [comment]
> Well, at the very least I think we should postpone this option until after we finally have something that's worthy of Blocks 0.1, and maybe it's something that could go in blocks-contrib instead.

I disagree. Filtering what is printed is basic functionality, and delaying its support until god knows when does seem be a good idea. Especially when a solution needs addition about 6 lines in Blocks (I am referring to the callback workaround).

> They all look nice, but they don't scale well at all. Once you've got 100 channels and want to match 50, you still end up writing a long list of patterns/regular expressions/etc. 

Right, but this is about making any filtering at all possible. And it would do the job in many cases, e.g. I am now printing norm and gradient magnitude for every parameters which I have about 20, and a super simple predicate would do the filtering job for me (`if record_name.startswith('/')`). 

I think your ncurses idea is nice, though it would require a good developer to make it near as flexible as the callback solution (e.g. to support interactive selection of a family of channels, etc.).

**Comment by bartvm** [comment]
> I disagree. Filtering what is printed is basic functionality, and delaying its support until god knows when does seem be a good idea.

There will always be reasonable proposals for new functionality. If we keep adding issues to the Blocks 0.1 milestone we'll never actually reach it (which is why it's a "god knows when" milestone in the first place). We also need to stop adding short-term workarounds, because they quickly become semi-permanent when other parts of the library start relying on them.

**Comment by rizar** [comment]
While I fully agree with what you say above, I think that when functionality is important enough, when it's not that short term workaround (in fact adding a callback is a universal solution which will finish development of `Printing` extension as far as I can see), when it does not require adding too many lines and when there is a person ready to contribute it, we should not lose these chances to make Blocks more usable.

In general I should say I not always like the line of code purity that you trying to enforce these days. Blocks have a couple of holes in which they do not deserve the nice 0.1 name, and this is one of them, and transition from coding fast and trying things to slow polishing of what we have should be gradual, taking into account user needs and opinions. As I apparently fail to convince you to take a more flexible stance, I think it's time to define Blocks management scheme.
